### ISSUE TITLE
Reverted changes comes back

### ISSUE THREAD

**Issue by grimmySwe** [ingydotnet/git-subrepo#108]
With the latest changes in 0.3.0 the current implementation is based on tree hashes for finding ancestors.

@jrosdahl Encountered a problem with this and there is a case where this fails:
Consider that you have a subrepo with tree hashes: `A->B->A`, in a main repo you have the tree hashes `A->B`. When you run pull you will actually find that the first subrepo A matches a previous A in the main repo, BUT you don't understand that it's actually connected to the previous A and the correct matching should be B to B. This pull will do nothing, treating the B as a valid change, when it should notice that B has been reverted into A.

**Comment by grimmySwe** [comment]
After a lot of thought I think that the only viable solution is to define possible common ancestors as totally unique tree hashes in both branches. So If you find a common tree hash it can't be found anywhere else, because then we can't determine if a revert of some kind has been in progress.

Solution is to fail in these cases and let the user sort out what he/she wants.

I'll continue tomorrow with implementation of a solution with this I will also consider #107 as it's in the same ball park.

**Comment by grimmySwe** [comment]
Implemented a solution for this on branch issue/107.

@ingydotnet During investigation of this issue I have consulted a lot with @jrosdahl (we do work at the same place :-) and there is some things to consider with this new tree hash solution.

It will be best where you interact or simply pull in. That way you will get a lot of common tree hashes.
It will be less good if you make a local change and never push it, that way you will never have common tree hashes and all operations are forced to use an older and older common ancestor. Worst case is that you only have the clone one in common.

The suggested workaround if you intend to have and keep local changes is to create a separate subrepo branch for that. And then perform the regular merges to that branch.

I have tried to come up with a solution to this but I see no easy way to do this.
